lib/software_version/version.rb

```diff
@@ -2,74 +2,154 @@ module SoftwareVersion
+    attr_reader :v,
+                :sv,
+                :epoch,
+                :version,
+                :revision,
+                :release,
+                :arch
+
+    def initialize(raw_version)
+      @v = raw_version
+      parse_raw_version(to_s)
+      parse_version
     end
 
     def <=>(other)
+      raise ArgumentError unless other.class == SoftwareVersion::Version
 
-      @sv.each_index do |k|
-        if (res = sub_compare(other, k)).nonzero?
-          return res
-        end
-      end
+      # Compare the epoch of both versions
+      result = version_compare_part(@epoch, other.epoch)
+      return result if result.nonzero?
+
+      # Compare the version of both versions
+      result = version_compare_part(@version, other.version)
+      return result if result.nonzero?
 
-      return @v <=> other.v if (@v <=> other.v).nonzero?
-      0
+      # Compare the revision of both versions
+      result = version_compare_part(@revision, other.revision)
+      return result if result.nonzero?
+
+      # Compare the release of both versions
+      version_compare_part(@release, other.release)
     end
 
     def to_s
-      @v
+      @v.to_s
     end
 
     def to_str
       to_s
     end
 
     def major
-      return nil if @sv.empty? || @sv[0].empty?
-      @sv[0][0]
+      return nil if @sv.empty?
+      @sv[0]
     end
 
     def minor
-      return nil if @sv.empty? || @sv[0].empty?
-      @sv[0][1]
+      return nil if @sv.empty?
+      @sv[1]
     end
 
     def patch
-      return nil if @sv.empty? || @sv[0].empty?
-      @sv[0][2]
+      return nil if @sv.empty?
+      @sv[2]
     end
 
-
     private
 
-    def sub_compare(other, k = 0)
-      # Split each sub part
-      @sv[k].each_with_index do |s, index|
-        # brek if the size of the array mismatch
-        break if other.sv[k].nil? || other.sv[k].size <= index
-
-        # check if the subpart is numeric
-        if numeric?(s) && numeric?(other.sv[k][index])
-          if (s.to_f <=> other.sv[k][index].to_f).nonzero?
-            return s.to_f <=> other.sv[k][index].to_f
-          else
-            next
-          end
+    # parse the raw version to separate the version, the epoch and the revision
+    def parse_raw_version(raw_version)
+      version = raw_version
+
+      if (parsed_raw = version.match(/^([^:]*):(.+)$/))
+        @epoch = parsed_raw[1]
+        version = parsed_raw[2]
+      else
+        @epoch = '0'
+      end
+
+      if (parsed_release = version.match(/(.*)\.(el[4-8](?:_\d+(?:\.\d+)?)?)/))
+        version = parsed_release[1]
+        @release = parsed_release[2]
+      else
+        @release = '0'
+      end
+
+      if (parsed_version = version.match(/(.*)-(.*)$/))
+        @version = parsed_version[1]
+        @revision = parsed_version[2]
+      else
+        @version = version
+        @revision = '0'
+      end
+    end
+
+    # Parse the version to get the major, minor and patch parts
+    def parse_version
```

Comment:
remove :sv attr_accesssor and define the accessor like that : 

```
def sv
  @sv ||= version.scan(/(?:[0-9]+|[A-Za-z]+)/)
end
```

remove parse_version from initializer.